sklearn/base.py

```diff
@@ -273,6 +273,8 @@ def __repr__(self, N_CHAR_MAX=700):
     def __getstate__(self):
         try:
             state = super().__getstate__()
+            if state is None:
+                state = self.__dict__.copy()
```

Comment:
One thing that puzzles me here is that [the docs say that `__getstate__()` will return `None](https://docs.python.org/3/library/pickle.html#object.__getstate__) if there is no instance `__dict__`. So how can we end up going down this `if` branch and then try to access `self.__dict__`? I've thought about this for a few minutes now but I can't work out when this would work, I'd expect that if we go down this path, then there is no `self.__dict__`, and so we can't access it here. So puzzled.

Reply:
Yes, I have been thoroughly confused by the documentation as well. What I think it doesn't mention, which is relevant here, is that if the object does not define `__getstate__` and if its instance dict is empty (but exists), it will also return None. This is the case we cover here.

```python
>>> sys.version
3.11.0
>>> class A:
...  pass
>>> print(A().__getstate__())
None
>>> A().__dict__
{}
```

Perhaps that's what they mean by:

> For a class that has no instance [`__dict__`](https://docs.python.org/3/library/stdtypes.html#object.__dict__) and no [`__slots__`](https://docs.python.org/3/reference/datamodel.html#object.__slots__), the default state is None.

but I'm not sure.

The other cases are indeed not covered, so this code can still break in very specific circumstances.

Reply:
I'm glad I'm not the only one puzzled by this. For me "has no instance `__dict__` sounds like "it doesn't exist" not "it exists but is empty". So yeah, confused.com

Could `state = self.__dict__.copy()` ever be anything than an empty dict? And if not, we could change it to `state = {}` no?

Reply:
I'm wondering about C extension types which don't necessarily have to comply with python object conventions.

This part of our code seems to be quite buggy if we consider all sorts of things people could do with `__dict__` and `__slots__` both on the instance and on the class (I know they shouldn't lol). But I'm okay with not supporting all those odd usecases.

Reply:
> Could state = self.__dict__.copy() ever be anything than an empty dict? And if not, we could change it to state = {} no?

I wondered about this too. Probably, it would always be an empty dict. If there is ever a situation (maybe Python 3.12?) where it's not an empty dict, would it be better to return that non-empty dict or `{}`? I thought the former, so I chose this solution. Not sure what other considerations to make here, runtime performance shouldn't matter much.

> But I'm okay with not supporting all those odd usecases.

At least as a first step, we could focus on making the existing sklearn estimators run again. I used the skops persistence tests, which should cover `__getstate__` of all estimators, and in addition to `KernelCenterer`, I also found `LabelEncoder`, which is a big one. Both of these don't have any C extensions, right?

There could be more estimators that I'm missing because they're masked by other errors in the test suite that occur due to Python 3.11.